src/v/v8_engine/instance.cc

```diff
@@ -0,0 +1,139 @@
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "v8_engine/instance.h"
+
+#include "utils/file_io.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/temporary_buffer.hh>
+#include <seastar/util/later.hh>
+
+#include <utility>
+
+namespace v8_engine {
+
+instance::instance(size_t max_heap_size_in_bytes) {
+    _create_params.array_buffer_allocator_shared
+      = std::shared_ptr<v8::ArrayBuffer::Allocator>(
+        v8::ArrayBuffer::Allocator::NewDefaultAllocator());
+    _create_params.constraints.ConfigureDefaultsFromHeapSize(
+      0, max_heap_size_in_bytes);
+    _isolate = v8::Isolate::New(_create_params);
+}
+
+instance::~instance() {
+    _function.Reset();
+    _context.Reset();
+    _isolate->Dispose();
+}
+
+ss::future<> instance::init(ss::sstring name, ss::sstring path) {
+    return compile_script(std::move(path)).then([this, name = std::move(name)] {
+        return get_function(name);
+    });
+}
+
+ss::future<> instance::compile_script(ss::sstring path) {
+    return read_fully_tmpbuf(path.data())
+      .then([this](const ss::temporary_buffer<char> script) mutable {
+          v8::Locker locker(_isolate);
+          v8::Isolate::Scope isolate_scope(_isolate);
+          v8::HandleScope handle_scope(_isolate);
+          v8::TryCatch try_catch(_isolate);
+          v8::Local<v8::Context> local_ctx = v8::Context::New(_isolate);
+          v8::Context::Scope context_scope(local_ctx);
+
+          v8::Local<v8::String> script_code = v8::String::NewFromUtf8(
+                                                _isolate,
+                                                script.begin(),
+                                                v8::NewStringType::kNormal,
+                                                script.size())
+                                                .ToLocalChecked();
+          v8::Local<v8::Script> compiled_script;
+          if (!v8::Script::Compile(local_ctx, script_code)
+                 .ToLocal(&compiled_script)) {
+              v8::String::Utf8Value error(_isolate, try_catch.Exception());
+              return ss::make_exception_future(exception(fmt::format(
+                "Can not compile script:{}",
+                std::string(*error, error.length()))));
+          }
+
+          v8::Local<v8::Value> result;
+          if (!compiled_script->Run(local_ctx).ToLocal(&result)) {
+              return ss::make_exception_future(
+                exception(fmt::format("Can not run script in first time")));
+          }
+
+          _context.Reset(_isolate, local_ctx);
+          return ss::now();
+      })
+      .handle_exception_type([](std::filesystem::filesystem_error& e) {
+          return ss::make_exception_future(
+            exception(fmt::format("Error: {}", e.what())));
+      });
+}
+
+ss::future<> instance::get_function(std::string_view name) {
```

Comment:
I guess this can return `void` instead of a `future`

Reply:
I used it to save exception from js code and write it in ss::future

Reply:
this is just a nit
if this code is only used with futures then I see no issue other than readability. the reader of the code might think that the code actually does something asynchronous and can be preempted but this isn't the case

Reply:
Done
It returns void